staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/values.go

```diff
@@ -0,0 +1,484 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"fmt"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/common/types/traits"
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	"k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model"
+	"reflect"
+	"sync"
+)
+
+// UnstructuredToVal converts a Kubernetes unstructured data element to a CEL Val.
+func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Structural) ref.Val {
```

Comment:
Question: Do you want this to be a part of your `cel.CustomTypeAdapter` or is this function used in a more narrow context?

Reply:
We don't have a CustomTypeAdapter in this code yet. We're currently only adding support for validation expressions with read-only access to the bound objects. What functionality/benefits do we get from integrating with CustomTypeAdapter? I'm generally in favor of integrating with cel-go properly.